src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncIteratorTests.cs

```diff
@@ -5012,5 +5012,782 @@ public async System.Collections.Generic.IAsyncEnumerator<int> GetAsyncEnumerator
+  IL_002a:  ldc.i4.s   -3
+  IL_002c:  newobj     ""C.<Iter>d__1..ctor(int)""
+  IL_0031:  stloc.0
+  IL_0032:  ldarga.s   V_1
+  IL_0034:  ldloca.s   V_1
+  IL_0036:  initobj    ""System.Threading.CancellationToken""
+  IL_003c:  ldloc.1
+  IL_003d:  call       ""bool System.Threading.CancellationToken.Equals(System.Threading.CancellationToken)""
+  IL_0042:  brfalse.s  IL_0052
+  IL_0044:  ldloc.0
+  IL_0045:  ldarg.0
+  IL_0046:  ldfld      ""System.Threading.CancellationToken C.<Iter>d__1.<>3__token1""
+  IL_004b:  stfld      ""System.Threading.CancellationToken C.<Iter>d__1.token1""
+  IL_0050:  br.s       IL_0059
+  IL_0052:  ldloc.0
+  IL_0053:  ldarg.1
+  IL_0054:  stfld      ""System.Threading.CancellationToken C.<Iter>d__1.token1""
+  IL_0059:  ldloc.0
+  IL_005a:  ldarg.0
+  IL_005b:  ldfld      ""int C.<Iter>d__1.<>3__value""
+  IL_0060:  stfld      ""int C.<Iter>d__1.value""
+  IL_0065:  ldloc.0
+  IL_0066:  ret
+}
+");
+            }
+        }
+
+        [Fact, WorkItem(34407, "https://github.com/dotnet/roslyn/issues/34407")]
+        public void CancellationTokenParameter_WrongParameterType()
+        {
+            string source = @"
+using System.Runtime.CompilerServices;
+using System.Threading.Tasks;
+class C
+{
+    static async System.Collections.Generic.IAsyncEnumerable<int> Iter([DefaultCancellation] int value)
+    {
+        yield return value++;
+        await Task.Yield();
+    }
+}";
+            var comp = CreateCompilationWithAsyncIterator(new[] { source, DefaultCancellationAttributeType });
+            comp.VerifyDiagnostics(
+                // (6,73): warning CS8424: The DefaultCancellationAttribute applied to parameter 'value' will have no effect. The attribute is only effective on a parameter of type CancellationToken in an async-enumerable method
+                //     static async System.Collections.Generic.IAsyncEnumerable<int> Iter([DefaultCancellation] int value)
+                Diagnostic(ErrorCode.WRN_UnconsumedDefaultCancellationAttributeUsage, "DefaultCancellation").WithArguments("value").WithLocation(6, 73)
+                );
+        }
+
+        [Fact, WorkItem(34407, "https://github.com/dotnet/roslyn/issues/34407")]
+        public void CancellationTokenParameter_AsyncEnumerator()
+        {
+            string source = @"
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+class C
+{
+    static async System.Collections.Generic.IAsyncEnumerator<int> Iter([DefaultCancellation] CancellationToken value)
+    {
+        yield return 1;
+        await Task.Yield();
+    }
+}";
+            var comp = CreateCompilationWithAsyncIterator(new[] { source, DefaultCancellationAttributeType });
+            comp.VerifyDiagnostics(
+                // (7,73): warning CS8424: The DefaultCancellationAttribute applied to parameter 'value' will have no effect. The attribute is only effective on a parameter of type CancellationToken in an async-enumerable method
```

Comment:
Consider having a different warning message for this case. My initial reaction here is "I'm writing an async enumerable and I have it on a parameter of type `CancellationToken`, is the compiler broken?

Reply:
It's subtle, but that's why I used "async enumerable" instead of "async iterator" (which could be either enumerable or enumerator). I can tweak the message to mention the IAsyncEnumerable type explicitly if that helps.